smdebug/tensorflow/keras.py

```diff
@@ -620,3 +623,86 @@ def new_add_weight(opt, *args, **kwargs):
         # Optimizer is being saved to support additional features in the future.
         self.optimizer = optimizer
         return optimizer
+
+    def forward_pre_hook(self, tape, mode=ModeKeys.GLOBAL):
+        """
+        Similar to _on_any_batch_begin(), this function prepares collections,
+        writes tensors, initializes writers, and increments step number
+        """
+        if self._is_not_supported():
+            return
+
+        self.tape = tape
+
+        self.worker = self._get_worker_name()
+        self.set_mode(mode)
+
+        if self.writer is not None or len(self.writer_map):
+            self._close_writers()
+
+        if not self.prepared_collections:
+            # at this point we need all collections to be ready
+            # this may not be the case at creation of hook
+            # as user's code after hook might add collections
+            self._prepare_collections()
+            self.prepared_collections = True
+
+        self._increment_step()
+
+        if self._get_collections_to_save_for_step():
+            self._initialize_writers()
+
+    def set_grads_vars(self, grads, vars):
```

Comment:
We could modify base_hook.set_gradients() to handle this case instead of introducing a new API.